File: app/src/main/java/com/kelsos/mbrc/commands/visual/VisualUpdateHandshakeComplete.java
package com.kelsos.mbrc.commands.visual;

import com.google.inject.Inject;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.data.SocketMessage;
import com.kelsos.mbrc.interfaces.ICommand;
import com.kelsos.mbrc.interfaces.IEvent;
import com.kelsos.mbrc.model.MainDataModel;
import com.kelsos.mbrc.services.SocketService;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import rx.Observable;

public class VisualUpdateHandshakeComplete implements ICommand {
  public static final String EMPTY = "";
  private SocketService service;
  private MainDataModel model;

  @Inject public VisualUpdateHandshakeComplete(SocketService service, MainDataModel model) {
    this.service = service;
    this.model = model;
  }

  public void execute(IEvent e) {
    boolean isComplete = (Boolean) e.getData();
    model.setHandShakeDone(isComplete);

    if (model.getPluginProtocol() < 2.1) {
      if (!isComplete) {
        return;
      }
      ArrayList<SocketMessage> messages = new ArrayList<>();
      messages.add(SocketMessage.create(Protocol.NowPlayingCover, EMPTY));
      messages.add(SocketMessage.create(Protocol.PlayerStatus, EMPTY));
      messages.add(SocketMessage.create(Protocol.NowPlayingTrack, EMPTY));
      messages.add(SocketMessage.create(Protocol.NowPlayingLyrics, EMPTY));
      messages.add(SocketMessage.create(Protocol.NowPlayingPosition, EMPTY));
      messages.add(SocketMessage.create(Protocol.PluginVersion, EMPTY));

      int totalMessages = messages.size();
      Observable.interval(150, TimeUnit.MILLISECONDS)
          .take(totalMessages)
          .subscribe(tick -> service.sendData(messages.remove(0)));
    } else {
      service.sendData(SocketMessage.create(Protocol.INIT, EMPTY));
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/commands/visual/VisualUpdateHandshakeComplete.kt
package com.kelsos.mbrc.commands.visual

import com.kelsos.mbrc.constants.Protocol
import com.kelsos.mbrc.data.SocketMessage
import com.kelsos.mbrc.interfaces.ICommand
import com.kelsos.mbrc.interfaces.IEvent
import com.kelsos.mbrc.model.ConnectionModel
import com.kelsos.mbrc.model.MainDataModel
import com.kelsos.mbrc.services.SocketService
import rx.Observable
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class VisualUpdateHandshakeComplete
@Inject constructor(private val service: SocketService, private val model: MainDataModel, private val connectionModel: ConnectionModel) : ICommand {

  override fun execute(e: IEvent) {
    val isComplete = e.data as Boolean
    connectionModel.setHandShakeDone(isComplete)

    if (!isComplete) {
      return
    }

    if (model.pluginProtocol > 2) {
      Timber.v("Sending init request")
      service.sendData(SocketMessage.create(Protocol.INIT))
    } else {

      Timber.v("Preparing to send requests for state")

      val messages = ArrayList<SocketMessage>()
      messages.add(SocketMessage.create(Protocol.NowPlayingCover))
      messages.add(SocketMessage.create(Protocol.PlayerStatus))
      messages.add(SocketMessage.create(Protocol.NowPlayingTrack))
      messages.add(SocketMessage.create(Protocol.NowPlayingLyrics))
      messages.add(SocketMessage.create(Protocol.NowPlayingPosition))
      messages.add(SocketMessage.create(Protocol.PluginVersion))

      val totalMessages = messages.size
      Observable.interval(150, TimeUnit.MILLISECONDS).take(totalMessages).subscribe { tick -> service.sendData(messages.removeAt(0)) }
    }
  }
}


